udica/man/man8/udica.8
.\" Copyright (C) 2018 Lukas Vrabec, <lvrabec@redhat.com>
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <https://www.gnu.org/licenses/>.

.\" A man page for udica.
.\" Contact lvrabec@redhat.com to report errors or typos.
.TH man 8 "17 February 2019" "1.1" "udica man page"

.SH NAME
udica \- a tool for generating SELinux security profiles for containers.

.SH SYNOPSIS
udica [options] container_name

.SH DESCRIPTION
A tool for generating SELinux security profiles for containers. The whole concept is based on the "block inheritance" feature inside the CIL intermediate language supported by the SELinux user space. The tool creates a policy which combines rules inherited from specified CIL blocks (templates) and rules discovered by inspection of the container JSON file with mount points and port definitions.

.SH SUPPORTED CONTAINER ENGINES
CRI-O v1.14.10+
docker v1.13+
podman v2.0+,

.SH OPTIONS
.TP
.I  \-h, \-\-help
Show this help message and exit

.TP
.I   \-i, \-\-container\-id ID
An ID of the running container for which should be an SELinux policy created

.TP
.I   \-j, \-\-json JSONFILE
Load JSON content of the inspected container from this file

.TP
.I   \-l, \-\-load\-modules
Load templates and module created by this tool

.TP
.I  \-c, \-\-caps CAPS
List of capabilities, for example: "\-c AUDIT\_WRITE,CHOWN,DAC\_OVERRIDE,FOWNER,FSETID,KILL"
(mandatory to use for Docker Engine, see the BUGS section)

.TP
.I  \-\-devices DEVS
List of devices the container should have access to, e.g "\-\-devices /dev/dri/card0,/dev/dri/renderD128"
(overrides devices specified in the conatiner JSON)

.TP
.I  \-a, \-\-append-rules FILE
Append more SELinux allow rules generated from SELinux denials in audit daemon.

.TP
.I  \-C, \-\-custom-rules FILE
Path to a CIL file with custom rules that will be included in the generated policy.
Rules must use 'process' as the source context to work correctly with the rest of the policy.
For example, to allow the container to connect to a unix stream socket:
.br
(allow process container_runtime_t ( unix_stream_socket ( connectto )))

.TP
.I  \-s, \-\-stream-connect DOMAIN
Allow container to stream connect with given SELinux domain.

.TP
.I \-e, \-\-container-engine ENGINE
Specify which container engine is used for the inspected container (supports: CRI-O, docker, podman)

.TP
.I   \-\-full\-network\-access
Allow a container full network access

.TP
.I   \-\-tty\-access
Allow a container to read and write the controlling terminal

.TP
.I   \-\-X\-access
Allow a container to communicate with the X server

.TP
.I   \-\-virt\-access
Allow a container to communicate with libvirt

.SH EXIT STATUS
.TP
.B 0
no errors encountered.
.TP
.B 1
other errors.
.TP
.B 2
error while parsing options.
.TP
.B 3
error while getting container info.
.TP
.B 4
error while creating an SELinux policy.

.SH EXAMPLES
.nf
# cat my_con.json | udica \-\-x\-access \-\-full\-network\-access my_container
Creates a new SELinux policy with name my_container based on inspecting container and adding access to the X server and full network access.

# echo "(allow process container_runtime_t ( unix_stream_socket ( connectto )))" > custom.cil
# udica \-j container.json \-\-custom\-rules custom.cil my_container
Creates a new SELinux policy module named my_container based on container.json and CIL rules from custom.cil.

.SH SEE ALSO
.BR selinux(8),
.BR podman(1)

.SH BUGS
Since it is not possible to detect capabilities used by a container in Docker Engine, you have to use '-c' to specify capabilities for docker container manually.

It is not possible to generate a custom local policy using the "audit2allow -M" command from AVCs where source context was generated by udica. For this purpose please use '--append-rules' option.

In some situations udica fails to identify which container engine is used, therefore "--container-engine" parameter has to be used to inform udica how JSON inspection file should be parsed.

.SH REPORTING BUGS
Report bugs to <https://github.com/containers/udica/issues/>

.SH AUTHOR
Written by Lukas Vrabec (lvrabec@redhat.com)


src/udica.py
#!/bin/python3

import sys
import subprocess
import argparse

import parse
import policy

def get_args():
    parser = argparse.ArgumentParser(description='Script generates SELinux policy for running container.')
    parser.add_argument(
        '-i', '--container-id', type=str, help='Running container ID', dest='ContainerID', required=True)
    parser.add_argument(
        '-n', '--name', type=str, help='Name for SELinux policy module', dest='ContainerName', required=True)
    parser.add_argument(
        '--full-network-access', help='Allow container full Network access ', required=False, dest='FullNetworkAccess', action='store_true')
    args = parser.parse_args()
    return vars(args)

def main():

    opts = get_args()

    container_inspect_data = subprocess.run(["podman", "inspect", opts['ContainerID']], capture_output=True).stdout.decode()
    container_caps_data = subprocess.run(["podman", "top", opts['ContainerID'], "capeff"], capture_output=True).stdout.decode()
    container_inspect = parse.parse_inspect(container_inspect_data)
    container_caps = parse.parse_cap(container_caps_data)

    container_mounts = container_inspect[0]['Mounts']
    container_ports = container_inspect[0]['NetworkSettings']['Ports']

    policy.create_policy(opts,container_caps,container_mounts,container_ports)

    print('\nPolicy ' + opts['ContainerName'] + ' with container id ' + opts['ContainerID'] + ' created!\n')
    print('Please load this module using: # semodule -i ' + opts['ContainerName'] + '.cil')
    print('Start container with: "--security-opt label=type:' + opts['ContainerName'] + '.process" parameter')

if __name__ == "__main__":
    main()
